openedx/core/djangoapps/user_authn/docs/decisions/0001_simplify_login_ratelimiting.rst

```diff
@@ -0,0 +1,95 @@
+
+Rate limiting Logins
+~~~~~~~~~~~~~~~~~~~~
+
+Currently there are five different rate limiting implementations in use as a
+part of the login flow.
+
+* `django-ratelimit`_ - A 3rd party rate limiting library that allows you to
+  decorate any view to add rate limiting functionality.
+
+  The ``ratelimit`` decorator on ``login_user`` uses the
+  LOGISTRATION_RATELIMIT_RATE django setting to determine the rate limit.  The
+  default is ``100 requests per 5 minutes``  This applies to both 1st and 3rd
+  party login attempts.
+
+* `django-ratelimit-backend`_ - Another 3rd party rate limiting library built
+  specifically to rate limit login views.  It provides a mixin that we can add
+  to an existing auth backend.  This replaces the base implementation of the
+  ``authenticate`` class with one that will also track the rate limits for that
+  call.
+
+  We currently use the default rate limits provided by the upstream library
+  which is ``30 requests per 5 minute``  This is only applies to 1st party
+  login attempts and does not count against 3rd party logins.
+
+* `djangorestframework rate limiting`_ - DRF provides a throttling
+  capability that can be used with any DRF views.  This is not used for the
+  login view but is applied to the ``/third_party_auth_context`` endpoint
+  (``TPAContextView``) to rate limit 3rd party auth requests.  The default
+  rate limit for this endpoint is ``20 requests per minute``.  Since this rate
+  limiting is only applied to a 3rd party auth specific view, it only impact
+  the rate at which 3rd party auth is possible.
+
+* `MAX_FAILED_LOGIN_ATTEMPTS`_ - This is an optional feature that can be
+  enable(disabled by default) which will limit the number of failed logins a
+  user is allowed to have before their account is locked out.  This feature
+  works slightly differently from the other rate limiting features in that it
+  persists the number of failures and does not reset them until we have had a
+  successful login.  If a maximum number of failed request is reached, the
+  account is locked out for 30 minutes.  The default settings for this feature
+  are to lock out the user for 30 minutes if 6 login failures occur over any
+  period of time.
+
+  edx.org has this feature enabled.
+
+* `cloudflare rate limiting`_ - This is edx.org specific and not enabled by
+  default for Open edX.
+
+.. _django-ratelimit: https://django-ratelimit.readthedocs.io/en/stable/usage.html#usage-chapter
+.. _django-ratelimit-backend: https://django-ratelimit-backend.readthedocs.io/en/latest/
+.. _djangorestframework rate limiting: https://www.django-rest-framework.org/api-guide/throttling/
+.. _MAX_FAILED_LOGIN_ATTEMPTS: https://github.com/edx/edx-platform/blob/cd6064692681ab99912e3da3721cd857a0b313e9/common/djangoapps/student/models.py#L980
+.. _cloudflare rate limiting: https://www.cloudflare.com/rate-limiting/
+
+Decisions
+---------
+
+* We will deprecate and remove the `django-ratelimit-backend`_ from
```

Comment:
Was there a specific set of reasons that led to this decision (vs deprecating/removing any of the other ones listed above), and if so, what were they?

Reply:
Agreed.  It would be nice to get more background info on the decisions.  It doesn't state that this library is no longer supported.  You also mentioned that some are simple to use, etc.